### PR TITLE
Add methods for converting dict to product or zip sweep and deprecate implicit cartesian product

### LINKED ISSUE #2698: Expand SWEEP_LIKE to work for {'t': [0, 1, 3]}
Currently if you make a call like `sampler.sample(circuit, params={'t': range(5)})` it doesn't work, because it expects e.g. a list of dictionaries instead of a dictionary containing lists. For convenience we should just support both cases. Check if the input is a dictionary with list values (and possibly singletons too), and if so take the cartesian product of the various entries and turn it into a sweep.

Test cases:

```
assert cirq.to_sweeps({'t': [0, 2, 3]}) == (
    cirq.to_sweeps([{'t': 0}, {'t': 2}, {'t': 3}]}))

assert cirq.to_sweeps({'t': [0, 1], 's': [2, 3], 'r': 4}) == (
    cirq.to_sweeps([
        {'t': 0, 's': 2, 'r': 4},
        {'t': 0, 's': 3, 'r': 4},
        {'t': 1, 's': 2, 'r': 4},
        {'t': 1, 's': 3, 'r': 4},
    ]))
```

### LINKED ISSUE #3018: Deprecate implicit cartesian product behavior in `to_sweeps`.
This one seems tricky to deprecate, because its arguments are often ultimately from a user but are provided by intermediate methods such as `cirq.sample`. This makes it difficult to print a relevant error message in terms of what the user did, instead of the internal details of the library.

### PR DESCRIPTION
**Pull request by kevinsung**
Fixes #2698 .
Fixes #3018 .

### CODE REVIEW

File: cirq/study/sweeps.py [hunk 1]
```diff
@@ -439,3 +435,43 @@ def _params_without_symbols(resolver: resolver.ParamResolver) -> Params:
         if isinstance(sym, sympy.Symbol):
             sym = sym.name
         yield cast(str, sym), cast(float, val)
+
+
+def dict_to_product_sweep(
+        factor_dict: Dict['cirq.TParamKey',
```
Comment (balopat) [inline cirq/study/sweeps.py:441]: I think having a named ZipOrProductSweepLike type alias would make this more readable, and signal it more clearly that the two methods are working off of the same type.

**Review by balopat — APPROVED** [review]
LGTM with one nitpick.

@crystalzhaizhai please have a look as well

File: cirq/study/sweeps.py (hunk 1 above)
Comment (kevinsung) [reply cirq/study/sweeps.py]: Done.

**Review by crystalzhaizhai — APPROVED** [review]
LGTM. 
quick question: why use cirq.TParamkey?

**Comment by kevinsung** [comment]
Parameter keys can be of type sympy.Basic in addition to string. We use TParamKey to represent the type of a parameter so we can define it in one place and also make the code more readable.

**Pull request merged**